lib/xwax/pitch_kalman.h

```diff
@@ -0,0 +1,325 @@
+
+    double P_pred[2][2];
+
+    P_pred[x][x] = F[x][x] * (F[x][x] * P[x][x] + F[x][v] * P[x][v]) +
+                   F[x][v] * (F[x][x] * P[x][v] + F[x][v] * P[v][v]) + Q[x][x];
+
+    P_pred[x][v] = F[v][x] * (F[x][x] * P[x][x] + F[x][v] * P[x][v]) +
+                   F[v][v] * (F[x][x] * P[x][v] + F[x][v] * P[v][v]) + Q[x][v];
+
+    P_pred[v][v] = F[v][x] * (F[v][x] * P[x][x] + F[v][v] * P[x][v]) +
+                   F[v][v] * (F[v][x] * P[x][v] + F[v][v] * P[v][v]) + Q[v][v];
+
+    /*
+     * =================== Update step =======================
+     *
+     *   Measurement model: z = Hx + noise
+     *   Here H = [1 0], so z = x + noise
+     *
+     *   Innovation (residual): y = z - H * x_pred
+     *   Since Hx = 1, Hv = 0, this reduces to: y = dx - x_pred
+     */
+
+    const double Hx = 1.0, Hv = 0.0;
+    const double y = dx - (Hx * x_pred + Hv * v_pred);
+
+    /*
+     * Toggle mode switches for stable playback and scratching. When the
+     * innovation quantity hits a certain treshold the filter sensitivity
+     * is tuned.
+     *
+     * The innovation is similar to the residual in the alpha-beta filter,
+     * calculated as
+     *
+     *          residual = measurement - prediction
+     *
+     * It represents how much the actual measurement deviates from the
+     * predicted value.
+     *
+     * NOTE: Probably better to use a continuous curve here instead of switches.
+     */
+
+    const double y_abs = fabs(y);
+
+    if (y_abs > p->scratch_threshold)
+        kalman_tune_sensitivity(p, &p->reactive);
+    else if (y_abs > p->medium_threshold && y_abs < p->scratch_threshold)
+        kalman_tune_sensitivity(p, &p->medium);
+    else
+        kalman_tune_sensitivity(p, &p->stable);
+
+    /*
+     * Innovation covariance: S = H * P_pred * H^T + R
+     * With H = [1 0], this reduces to: S = P_pred[x][x] + R
+     */
+
+    const double S = P_pred[x][x] + p->coeffs->R;
+
+    /*
+     * Kalman gain: K = P_pred * H^T * S^-1
+     *
+     * With H = [1 0]:
+     *   Kx = P_pred[x][x] / S
+     *   Kv = P_pred[x][v] / S
+     */
+
+    const double invS = 1.0 / S;
+    const double Kx = P_pred[x][x] * invS;
+    const double Kv = P_pred[x][v] * invS;
+
+    /*
+     * Update state:
+     *
+     * x_upd = x_pred + Kx * y
+     * v_upd = v_pred + Kv * y
+     */
+
+    double x_upd = x_pred + Kx * y;
+    const double v_upd = v_pred + Kv * y;
+
+    /*
+     * Update covariance: P = (I - K H) * P_pred
+     *
+     * H = [1 0], so:
+     *   I - K H = [ 1-Kx     0
+     *              -Kv       1 ]
+     *
+     * Then:
+     *   P_upd[x][x] = (1-Kx) * P_pred[x][x]
+     *   P_upd[x][v] = (1-Kx) * P_pred[x][v]   (same as Pvx_upd)
+     *   P_upd[v][v] = P_pred[v][v] - Kv * P_pred[x][v]
+     */
+
+    double P_upd[2][2];
+
+    P_upd[x][x] = (1.0 - Kx) * P_pred[x][x];
+    P_upd[x][v] = (1.0 - Kx) * P_pred[x][v]; /* equals Pvx_upd */
+    P_upd[v][v] = P_pred[v][v] - Kv * P_pred[x][v];
+
```

Comment:
same partial initilization issue here IMO.

Reply:
It's a draft